File: src/main/java/com/kishlaly/ta/utils/Dates.java
package com.kishlaly.ta.utils;

import com.kishlaly.ta.model.Quote;
import com.kishlaly.ta.model.Timeframe;

import java.time.*;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;

/**
 * @author Vladimir Kishlaly
 * @since 19.11.2021
 */
public class Dates {

    public static String beautifyQuoteDate(Quote quote) {
        return getBarTimeInMyZone(quote.getTimestamp(), quote.exchangeTimezome).toString();
    }

    public static String getDateFormat() {
        switch (Context.timeframe) {
            case HOUR:
                return "yyyy-MM-dd HH:mm:ss";
            case WEEK:
            default:
                return "yyyy-MM-dd";
        }
    }

    // on the input may be the date of the form yyyyy-MM-dd for the day timeframe, so I shift the clock to the beginning of the exchange
    public static ZonedDateTime getTimeInExchangeZone(String date, String exchangeTimezone) {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern(getDateFormat());
        switch (Context.timeframe) {
            case HOUR:
                LocalDateTime localDate = LocalDateTime.parse(date, formatter);
                return localDate
                        .atZone(ZoneId.of(exchangeTimezone))
                        .withZoneSameInstant(ZoneId.of(exchangeTimezone));
            case WEEK:
            case DAY:
            default:
                return LocalDate.parse(date, formatter)
                        .atStartOfDay(ZoneId.of(exchangeTimezone))
                        .withZoneSameInstant(ZoneId.of(exchangeTimezone))
                        .plus(9, ChronoUnit.HOURS)
                        .plus(30, ChronoUnit.MINUTES);
        }
    }

    // on the input may be the date of the form yyyyy-MM-dd for the day timeframe, so I shift the clock to the beginning of the exchange
    public static ZonedDateTime getBarTimeInMyZone(String date, String exchangeTimezone) {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern(getDateFormat());
        switch (Context.timeframe) {
            case HOUR:
                LocalDateTime localDate = LocalDateTime.parse(date, formatter);
                return localDate
                        .atZone(ZoneId.of(exchangeTimezone))
                        .withZoneSameInstant(ZoneId.of(Context.myTimezone));
            case WEEK:
            case DAY:
            default:
                return LocalDate.parse(date, formatter)
                        .atStartOfDay(ZoneId.of(exchangeTimezone))
                        .withZoneSameInstant(ZoneId.of(Context.myTimezone))
                        .plus(9, ChronoUnit.HOURS)
                        .plus(30, ChronoUnit.MINUTES);
        }
    }

    // At the entrance of the exact time, so there is no need to reset the clock at the time of the opening of the exchange
    public static ZonedDateTime getBarTimeInMyZone(Long timestamp, String exchangeTimezoneStr) {
        ZoneId exchangeTimezone = ZoneId.of(exchangeTimezoneStr);
        ZoneId myTimezone = ZoneId.of(Context.myTimezone);

        LocalDateTime exchangeLocalDateTime = LocalDateTime.ofInstant(Instant.ofEpochSecond(timestamp), exchangeTimezone);
        ZonedDateTime exchangeZonedDateTime = exchangeLocalDateTime.atZone(exchangeTimezone);

        ZonedDateTime timeInMyZone = exchangeZonedDateTime.withZoneSameInstant(OffsetDateTime.now(myTimezone).getOffset()).withZoneSameInstant(myTimezone);
        return timeInMyZone;
    }

    public static ZonedDateTime getTimeInExchangeZone(Long timestamp, String exchangeTimezoneStr) {
        ZoneId exchangeTimezone = ZoneId.of(exchangeTimezoneStr);
        LocalDateTime exchangeLocalDateTime = LocalDateTime.ofInstant(Instant.ofEpochSecond(timestamp), exchangeTimezone);
        ZonedDateTime exchangeZonedDateTime = exchangeLocalDateTime.atZone(exchangeTimezone);
        return exchangeZonedDateTime;
    }

    public static String getDuration(Timeframe timeframe, long start, long end) {
        ZoneId timezone = ZoneId.of(Context.myTimezone);
        LocalDateTime startDate = LocalDateTime.ofInstant(Instant.ofEpochSecond(start), timezone);
        LocalDateTime endDate = LocalDateTime.ofInstant(Instant.ofEpochSecond(end), timezone);
        switch (timeframe) {
            case WEEK:
                return ChronoUnit.WEEKS.between(startDate, endDate) + " weeks";
            case DAY:
                return ChronoUnit.DAYS.between(startDate, endDate) + " days";
            case HOUR:
                return ChronoUnit.HOURS.between(startDate, endDate) + " hours";
            default:
                return "";
        }
    }

    public static ZonedDateTime shortDateToZoned(String datePart) {
        String[] split = datePart.split("\\.");
        String date = split[2] + "-" + split[1] + "-" + split[0] + "T09:30-04:00[US/Eastern]";
        ZonedDateTime parsed = ZonedDateTime.parse(date);
        return parsed;
    }

    public static void main(String[] args) {
        System.out.println(getDuration(Timeframe.DAY, 1632749400, 1633008600));
    }

}

File: src/main/kotlin/com/kishlaly/ta/utils/Dates.kt
package com.kishlaly.ta.utils

import com.kishlaly.ta.config.Context
import com.kishlaly.ta.model.Quote
import com.kishlaly.ta.model.Timeframe
import com.kishlaly.ta.model.exchangeTimezome
import java.time.*
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit

class Dates {

    companion object {

        fun beautifyQuoteDate(quote: Quote): String {
            return getBarTimeInMyZone(quote.timestamp, exchangeTimezome).toString()
        }

        fun getDateFormat(): String {
            return when (Context.timeframe) {
                Timeframe.HOUR -> "yyyy-MM-dd HH:mm:ss"
                Timeframe.WEEK -> "yyyy-MM-dd"
                else -> "yyyy-MM-dd"
            }
        }

        // on the input may be the date of the form yyyyy-MM-dd for the day timeframe, so I shift the clock to the beginning of the exchange
        fun getTimeInExchangeZone(date: String, exchangeTimezone: String): ZonedDateTime {
            val formatter = DateTimeFormatter.ofPattern(getDateFormat())
            return when (Context.timeframe) {
                Timeframe.HOUR -> {
                    val localDate = LocalDateTime.parse(date, formatter)
                    localDate
                            .atZone(ZoneId.of(exchangeTimezone))
                            .withZoneSameInstant(ZoneId.of(exchangeTimezone))
                }
                Timeframe.WEEK, Timeframe.DAY -> LocalDate.parse(date, formatter)
                        .atStartOfDay(ZoneId.of(exchangeTimezone))
                        .withZoneSameInstant(ZoneId.of(exchangeTimezone))
                        .plus(9, ChronoUnit.HOURS)
                        .plus(30, ChronoUnit.MINUTES)
            }
        }

        // on the input may be the date of the form yyyyy-MM-dd for the day timeframe, so I shift the clock to the beginning of the exchange
        fun getBarTimeInMyZone(date: String, exchangeTimezone: String): ZonedDateTime {
            val formatter = DateTimeFormatter.ofPattern(getDateFormat())
            return when (Context.timeframe) {
                Timeframe.HOUR -> {
                    val localDate = LocalDateTime.parse(date, formatter)
                    localDate
                            .atZone(ZoneId.of(exchangeTimezone))
                            .withZoneSameInstant(ZoneId.of(Context.myTimezone))
                }
                Timeframe.WEEK, Timeframe.DAY -> LocalDate.parse(date, formatter)
                        .atStartOfDay(ZoneId.of(exchangeTimezone))
                        .withZoneSameInstant(ZoneId.of(Context.myTimezone))
                        .plus(9, ChronoUnit.HOURS)
                        .plus(30, ChronoUnit.MINUTES)
            }
        }

        // At the entrance of the exact time, so there is no need to reset the clock at the time of the opening of the exchange
        fun getBarTimeInMyZone(timestamp: Long, exchangeTimezoneStr: String): ZonedDateTime {
            val exchangeTimezone = ZoneId.of(exchangeTimezoneStr)
            val myTimezone = ZoneId.of(Context.myTimezone)
            val exchangeLocalDateTime = LocalDateTime.ofInstant(Instant.ofEpochSecond(timestamp), exchangeTimezone)
            val exchangeZonedDateTime = exchangeLocalDateTime.atZone(exchangeTimezone)
            return exchangeZonedDateTime.withZoneSameInstant(OffsetDateTime.now(myTimezone).offset).withZoneSameInstant(myTimezone)
        }

        fun getTimeInExchangeZone(timestamp: Long, exchangeTimezoneStr: String): ZonedDateTime {
            val exchangeTimezone = ZoneId.of(exchangeTimezoneStr)
            val exchangeLocalDateTime = LocalDateTime.ofInstant(Instant.ofEpochSecond(timestamp!!), exchangeTimezone)
            return exchangeLocalDateTime.atZone(exchangeTimezone)
        }

        fun getDuration(timeframe: Timeframe, start: Long, end: Long): String {
            val timezone = ZoneId.of(Context.myTimezone)
            val startDate = LocalDateTime.ofInstant(Instant.ofEpochSecond(start), timezone)
            val endDate = LocalDateTime.ofInstant(Instant.ofEpochSecond(end), timezone)
            return when (timeframe) {
                Timeframe.WEEK -> ChronoUnit.WEEKS.between(startDate, endDate).toString() + " weeks"
                Timeframe.DAY -> ChronoUnit.DAYS.between(startDate, endDate).toString() + " days"
                Timeframe.HOUR -> ChronoUnit.HOURS.between(startDate, endDate).toString() + " hours"
            }
        }

        fun shortDateToZoned(datePart: String): ZonedDateTime {
            val split = datePart.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            val date = split[2] + "-" + split[1] + "-" + split[0] + "T09:30-04:00[US/Eastern]"
            return ZonedDateTime.parse(date)
        }
    }

}